qubes/vm/qubesvm.py

```diff
@@ -695,6 +683,49 @@ def __xml__(self):
 
         return element
 
+    def libvirt_connected(self):
+        self._update_libvirt_state()
+
+    def libvirt_lifecycle_event(self, event, detail):
+        # TODO: we could perhaps determine the new state from the event
+        #   without asking libvirtd
+        self._update_libvirt_state()
+
+    def _update_libvirt_state(self):
+        was_halted = self.is_halted()
+
+        if self.libvirt_domain is not None:
+            try:
+                state = self.libvirt_domain.state()
+                self._libvirt_state = state[0]
+                self._libvirt_substate = state[1]
+            except libvirt.libvirtError as e:
+                if e.get_error_code() != libvirt.VIR_ERR_NO_DOMAIN:
+                    self.log.exception('libvirt error code: {!r}'.format(
+                        e.get_error_code()))
+                    raise
+
+            try:
+                self._xid = self.libvirt_domain.ID()
+            except libvirt.libvirtError as e:
+                if e.get_error_code() != libvirt.VIR_ERR_NO_DOMAIN:
+                    self.log.exception('libvirt error code: {!r}'.format(
+                        e.get_error_code()))
+                    raise
+
+            if self.is_running() and self.app.vmm.xs is not None:
+                stubdom_xid_str = self.app.vmm.xs.read('',
+                    '/local/domain/{}/image/device-model-domid'
+                        .format(self.xid))
+                if stubdom_xid_str is not None and stubdom_xid_str.isdigit():
+                    self._stubdom_xid = int(stubdom_xid_str)
+
+        if self.is_halted() and not was_halted:
+            self.fire_event('domain-shutdown')
```

Comment:
This prevents the "double shutdown event" as well. It think just taking the startup lock while firing the event could fix the synchronization issue?

Reply:
It will not prevent handling domain-shutdown after VM was started again, namely this sequence:
1. VM shutdown
2. vm.start()
3. vm.fire_event('domain-shutdown')

You don't have any guarantee about order of input processing (request on qubesd.sock, event on libvirt socket). And additionally, you can't take lock directly in libvirt event handler, because it isn't coroutine.

Reply:
So I think this needs to be thought about and done more carefully: probably the simplest way is having non-coroutines schedule coroutines, and then pretty much serializing all methods that change the VM in any way, possibly including libvirt state updates, by taking the startup_lock (which should just be renamed to "lock") and updating libvirt state under lock if necessary.

Could also consider deleting the libvirt domain for VMs that are shutdown to prevent external tools starting it while we finalize storage and other such robustness measures against raw use of the libvirt/xen APIs.

I think neither the current code nor this patch does this correctly, and some more work is needed.